<!DOCTYPE html>
<html>
<body>

<h2>JavaScript Error Handling</h2>
<p>How to use <b>catch</b> to display an error.</p>
<p id="demo1"></p>

<script>
    try {
        adddlert("Welcome guest!");
    } catch (err) {
        document.getElementById("demo1").innerHTML = err.message;
    }
</script>


<h2>JavaScript try catch</h2>
<p>Please input a number between 5 and 10:</p>
<input id="demo2" type="text">
<button type="button" onclick="myFunctionA()">Test Input</button>
<p id="p01"></p>

<script>
    function myFunctionA() {
        const message = document.getElementById("p01");
        message.innerHTML = "";
        let x = document.getElementById("demo2").value;
        try {
            if (x.trim() == "") throw "empty";
            if (isNaN(x)) throw "not a number";
            x = Number(x);
            if (x < 5) throw "too low";
            if (x > 10) throw "too high";
        } catch (err) {
            message.innerHTML = "Input is " + err;
        }
    }
</script>


<h2>JavaScript try catch</h2>
<p>Please input a number between 5 and 10:</p>
<input id="demo3" type="text">
<button type="button" onclick="myFunctionB()">Test Input</button>
<p id="p02"></p>

<script>
    function myFunctionB() {
        const message = document.getElementById("p02");
        message.innerHTML = "";
        let xA = document.getElementById("demo3").value;
        try {
            if (xA.trim() == "") throw "is empty";
            if (isNaN(xA)) throw "is not a number";
            xA = Number(xA);
            if (xA > 10) throw "is too high";
            if (xA < 5) throw "is too low";
        } catch (err) {
            message.innerHTML = "Input " + err;
        } finally {
            document.getElementById("demo3").value = "";
        }
    }
</script>


<h2>JavaScript Errors</h2>
<p>You cannot set the number of significant digits of a number to 500:</p>
<p id="demo4">

    <script>
        let num = 1;
        try {
            num.toPrecision(500);
        } catch (err) {
            document.getElementById("demo4").innerHTML = err.name;
        }
    </script>


<h2>JavaScript Errors</h2>
<p>You cannot use the value of a non-existing variable:</p>
<p id="demo5"></p>

<script>
    let x = 5;
    try {
        x = y + 1;
    } catch (err) {
        document.getElementById("demo5").innerHTML = err.name;
    }
</script>


<h2>JavaScript Errors</h2>
<p>You cannot evaluate code that contains a syntax error:</p>
<p id="demo6"></p>

<script>
    try {
        eval("alert('Hello')");
    } catch (err) {
        document.getElementById("demo6").innerHTML = err.name;
    }
</script>


<h2>JavaScript Errors</h2>
<p>You cannot convert a number to upper case:</p>
<p id="demo7"></p>

<script>
    let numA = 1;
    try {
        num.toUpperCase();
    } catch (err) {
        document.getElementById("demo7").innerHTML = err.name;
    }
</script>


<h2>JavaScript Errors</h2>
<p>Some characters cannot be decoded with decodeURI():</p>
<p id="demo8"></p>

<script>
    try {
        decodeURI("%%%");
    } catch (err) {
        document.getElementById("demo8").innerHTML = err.name;
    }
</script>
</body>
</html>